### PR TITLE
[TACHYON-1576] Force each key-value partition no larger than the max size set in Conf

### PR DESCRIPTION
**Pull request by apc999**
https://tachyon.atlassian.net/browse/TACHYON-1576

### CODE REVIEW

**Comment by apc999** [comment]
Jenkins, test this again.

File: keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java [hunk 1]
```diff
@@ -102,14 +102,15 @@ public void put(byte[] key, byte[] value) throws IOException {
...
     Preconditions.checkState(!mClosed);
-    Preconditions.checkState(!isFull());
     mIndex.put(key, value, mPayloadWriter);
     mKeyCount ++;
   }
 
   @Override
-  public boolean isFull() {
-    return byteCount() >= mMaxSizeBytes;
+  public boolean canPutKeyValue(byte[] key, byte[] value) {
```
Comment (jsimsa) [inline keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java]: just canPut? KeyValue seems redundant

**Comment by jsimsa** [comment]
just one minor comment

File: integration-tests/src/test/java/tachyon/client/keyvalue/KeyValueStoresIntegrationTest.java [hunk 2]
```diff
@@ -118,4 +118,44 @@ public void createAndOpenStoreWithMultiKeysTest() throws Exception {
...
+
+  /**
+   * Tests creating and opening a store with a number of keys, while each key-value pair is large
+   * enough to take a separate key-value partition.
+   */
+  @Test
+  public void createMultiPartitionsTest() throws Exception {
+    final long maxPartitionSize = Constants.MB; // Each partition is at most 1 MB
+    final int numKeys = 10;
+    final int keyLength = 4; // 64 Byte key
```
Comment (uronce-cc) [inline integration-tests/src/test/java/tachyon/client/keyvalue/KeyValueStoresIntegrationTest.java]: `64 Byte key`? seems like it is a `4 Byte key`?

File: keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java [hunk 3]
```diff
@@ -102,14 +102,15 @@ public void put(byte[] key, byte[] value) throws IOException {
...
+    // See BasePayloadWriter.insert()
+    return byteCount() + key.length + value.length
```
Comment (uronce-cc) [inline keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java]: We didn't estimate size increase in Index after putting the kv pair?

**Comment by apc999** [comment]
Jenkins, test this please.

File: keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java (hunk 1 above)
Comment (apc999) [inline keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java]: done

File: integration-tests/src/test/java/tachyon/client/keyvalue/KeyValueStoresIntegrationTest.java (hunk 2 above)
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/keyvalue/KeyValueStoresIntegrationTest.java]: fixed

File: keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java (hunk 3 above)
Comment (apc999) [inline keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValuePartitionWriter.java]: for now, no estimation as index growth is not supported yet.
A TODO left.

**Comment by uronce-cc** [comment]
LGTM

**Comment by apc999** [comment]
@jsimsa , I fixed the comment. PTAL

**Comment by jsimsa** [comment]
LGTM

**Pull request merged**